### PR TITLE
Clear out ClientCache when UI is detached to prevent a minor memory leak

### PR DESCRIPTION
**Pull request by Ansku**
Implemented with a listener rather than direct call from UI.detach() in
order to avoid new public API, since the whole feature has been marked
for removal. This doesn't yet prevent the cache or the type map from
getting slightly bloated during the UI's lifetime.

See: #3705

Cherry-picked from: #12199

---
This change is [Reviewable](https://reviewable.io/reviews/vaadin/framework/12200)

### CODE REVIEW

**Review by TatuLund — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/vaadin/framework/12200#-:-MTPU1fDEPrTS78TRz2I:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved
